src/wallet.rs

```diff
@@ -0,0 +1,345 @@
+		let tokio_runtime = RwLock::new(None);
+		Self { blockchain, inner, fee_rate_cache, tokio_runtime, logger }
+	}
+
+	pub(crate) async fn sync(&self) -> Result<(), Error> {
+		match self.update_fee_estimates().await {
+			Ok(()) => (),
+			Err(e) => {
+				log_error!(self.logger, "Fee estimation error: {}", e);
+				return Err(e);
+			}
+		}
+
+		let sync_options = SyncOptions { progress: None };
+		match self.inner.lock().unwrap().sync(&self.blockchain, sync_options).await {
+			Ok(()) => Ok(()),
+			Err(e) => {
+				log_error!(self.logger, "Wallet sync error: {}", e);
+				Err(From::from(e))
+			}
+		}
+	}
+
+	pub(crate) fn set_runtime(&self, tokio_runtime: Arc<tokio::runtime::Runtime>) {
+		*self.tokio_runtime.write().unwrap() = Some(tokio_runtime);
+	}
+
+	pub(crate) fn drop_runtime(&self) {
+		*self.tokio_runtime.write().unwrap() = None;
+	}
+
+	pub(crate) async fn update_fee_estimates(&self) -> Result<(), Error> {
+		let mut locked_fee_rate_cache = self.fee_rate_cache.write().unwrap();
+
+		let confirmation_targets = vec![
+			ConfirmationTarget::Background,
+			ConfirmationTarget::Normal,
+			ConfirmationTarget::HighPriority,
+		];
+		for target in confirmation_targets {
+			let num_blocks = match target {
+				ConfirmationTarget::Background => 12,
+				ConfirmationTarget::Normal => 6,
+				ConfirmationTarget::HighPriority => 3,
+			};
+
+			let est_fee_rate = self.blockchain.estimate_fee(num_blocks).await;
+
+			match est_fee_rate {
+				Ok(rate) => {
+					locked_fee_rate_cache.insert(target, rate);
+					log_trace!(
+						self.logger,
+						"Fee rate estimation updated: {} sats/kwu",
+						rate.fee_wu(1000)
+					);
+				}
+				Err(e) => {
+					log_error!(self.logger, "Failed to update fee rate estimation: {}", e);
+				}
+			}
+		}
+		Ok(())
+	}
+
+	pub(crate) fn create_funding_transaction(
+		&self, output_script: Script, value_sats: u64, confirmation_target: ConfirmationTarget,
+	) -> Result<Transaction, Error> {
+		let fee_rate = self.estimate_fee_rate(confirmation_target);
+
+		let locked_wallet = self.inner.lock().unwrap();
+		let mut tx_builder = locked_wallet.build_tx();
+
+		tx_builder.add_recipient(output_script, value_sats).fee_rate(fee_rate).enable_rbf();
+
+		let mut psbt = match tx_builder.finish() {
+			Ok((psbt, _)) => {
+				log_trace!(self.logger, "Created funding PSBT: {:?}", psbt);
+				psbt
+			}
+			Err(err) => {
+				log_error!(self.logger, "Failed to create funding transaction: {}", err);
+				Err(err)?
+			}
+		};
+
+		// We double-check that no inputs try to spend non-witness outputs. As we use a SegWit
+		// wallet descriptor this technically shouldn't ever happen, but better safe than sorry.
+		for input in &psbt.inputs {
+			if input.witness_utxo.is_none() {
+				log_error!(self.logger, "Tried to spend a non-witness funding output. This must not ever happen. Panicking!");
+				panic!("Tried to spend a non-witness funding output. This must not ever happen.");
+			}
+		}
```

Comment:
FYI this check is redundant because LDK already checks for this internally:

https://github.com/lightningdevkit/rust-lightning/blob/01fe9badc034fdc16a4450a2ce9ba5552560329f/lightning/src/ln/channelmanager.rs#L2644-L2650

It could still be kept around though, up to you

Reply:
Ah, right, good point. Now removed it as this would be triply redundant (as we use Bip84 wallet descriptor, this shouldn't ever happen anyways).